File: Caper/CaperSharp/Sequence.cs
using System;
using System.IO;
using System.Text;
using System.Text.RegularExpressions;

namespace CaperSharp
{
  public class Sequence
  {
    private const char NewLine = '\n';

    private string mBases;

    private StreamReader mHandle;
    private long mPosition;
    public long Length { get; private set; }

    private bool mIsStream;

    internal Sequence(string aBases)
    {
      mBases = aBases;
      mIsStream = false;
      Length = aBases.Length;
    }

    internal Sequence(Stream aHandle, long aPosition, long aLength)
    {
      mHandle = new StreamReader(aHandle);
      mPosition = aPosition;
      Length = aLength;

      mIsStream = true;
    }

    internal char this[int aIndex]
    {
      get { return mBases[aIndex]; }
    }

    internal string Substring(long aIndex, long aCount)
    {
      if (!mIsStream)
        return mBases.Substring((int)aIndex, (int)aCount);
      else
      {
        if (aCount > Length)
          throw new Exception("Could not seek as far as requested.");

        mHandle.BaseStream.Seek(mPosition + aIndex, SeekOrigin.Begin);
        mHandle.DiscardBufferedData();
        char[] lBuffer = new char[aCount];
        if (mHandle.Read(lBuffer, 0, (int)aCount) < aCount)
          throw new Exception("Could not read sequence from handle.");

        string lBufferedString = new string(lBuffer);

        return Regex.Replace(lBufferedString, "\n", "");
      }
    }

    public override string ToString()
    {
      return Substring(0, Length);

    }
  }
}

File: Caper/Caper/Sequence.cpp
#include "Sequence.h"

Sequence::Sequence( string aBases )
{
	mBases = aBases;
  mPath = "";
  mIsStream = false;
  Length = aBases.length();
}

Sequence::Sequence( string & aPath, long long & aPosition, long long & aLength )
{
  mBases = "";
  mHandle.open( aPath.c_str() );
  mPosition = aPosition;
  Length = aLength;

  mIsStream = true;
}

Sequence::Sequence( Sequence & aSequence )
{
  mBases = aSequence.mBases;
  mPath = aSequence.mPath;

  mPosition = aSequence.mPosition;
  Length = aSequence.Length;

  mIsStream = aSequence.mIsStream;

  if ( aSequence.mIsStream )
    mHandle.open( aSequence.mPath.c_str() );
}

string Sequence::Substring( long long aIndex )
{
  return Substring( aIndex, Length - aIndex );
}

string Sequence::Substring( long long aIndex, long long aCount )
{
  if ( !mIsStream )
    return mBases.substr((int) aIndex, (int) aCount ); // TODO, could create a nasty overflow
  else
  {
    if ( aCount > Length )
      throw string( "could not seek as far as requested" );

    mHandle.seekg( mPosition + aIndex );
    char * lBuffer = new char[aCount + 1];
    mHandle.get( lBuffer, aCount );
    lBuffer[aCount] = 0; // add the null char to terminate the string

    int lCount = mHandle.gcount();

    if ( !mHandle.good() )
      throw string("could not read whole sequence from handle.");
  
    return string(lBuffer);
  }
}

string Sequence::ToString()
{
  long long lCount =0;
  return Substring(lCount, Length);
}

Sequence::~Sequence(void)
{
  mHandle.close();
}
